python/lsst/pipe/base/quantum_provenance_graph.py

```diff
@@ -1075,6 +1126,110 @@ def make_exception_table(self) -> astropy.table.Table:
                 rows.append({"Task": task_label, "Exception": type_name, "Count": len(exception_summaries)})
         return astropy.table.Table(rows)
 
+    def make_exception_diagnostics_table(
+        self,
+        butler: Butler | None = None,
+        add_exception_msg: bool = True,
+        max_message_width: int | None = None,
+        shorten_type_name: bool = False,
+    ) -> astropy.table.Table:
+        """Construct an `astropy.table.Table` showing exceptions grouped by
+        data ID.
+
+        Each row represents one data ID that encountered an exception, along
+        with the exception type under the column named after the task that
+        raised it. If a Butler is provided, the table will also include a
+        subset of exposure-related metadata pulled from the exposure dimension
+        records. The exception message can optionaly be included in the table.
+
+        Parameters
+        ----------
+        butler : `lsst.daf.butler.Butler`, optional
+            Butler instance used to fetch exposure records. If not provided,
+            exposure dimension records will not be included in the table.
+        add_exception_msg : `bool`, optional
+            If `True`, include the exception message in the table.
+        max_message_width : `int`, optional
+            Maximum width for storing exception messages in the output table.
+            Longer messages will be truncated. If not provided, messages will
+            be included in full without truncation.
+        shorten_type_name : `bool`, optional
+            If `True`, shorten the exception type name by removing the
+            package name. This is useful for making the table more readable
+            when the package name is long or not relevant to the user.
+
+        Returns
+        -------
+        table : `astropy.table.Table`
+            Table with one row per data ID and columns for exception types (by
+            task), and optionally, exposure dimension records and exception
+            messages.
+        """
+        add_exposure_records = True
+        needed_exposure_records = ["day_obs", "physical_filter", "exposure_time", "target_name"]
+
+        # Preload all exposure dimension records up front for faster O(1)
+        # lookup later. Querying per data ID in the loop is painfully slow.
+        if butler:
+            exposure_record_lookup = {
+                d.dataId["exposure"]: d for d in butler.query_dimension_records("exposure", explain=False)
```

Comment:
I wish I could limit `butler.query_dimension_records` to just the data IDs with exceptions, instead of pulling everything. But it only takes one dataId, and I don't think it's possible to use `where` to filter it.

Reply:
This works fine for me if you already know the exposure IDs:
```python
>>> butler.query_dimension_records("exposure", where="exposure in (:exp)", bind={"exp": (903334, 903336)}, instrument="HSC")
```

Reply:
Use
```
with butler.query() as query:
    query = query.join_data_coordinates(<list of exposure data IDs>)
    ...
```

Reply:
Actually, change `.required` to `.mapping` in this line:

https://github.com/lsst/pipe_base/blob/924f49135fb97c9293592500239795adae769306/python/lsst/pipe/base/quantum_provenance_graph.py#L658

and then the data IDs you're starting with should have all of those implied values.  They're already in the data IDs in the QG, but we were dropping them on the floor.

Reply:
Will give it a try!

Reply:
Done!